pkg/sdkclient/batchmapper/client.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		return nil, err
+	}
+
+	c := new(client)
+	c.conn = conn
+	c.grpcClt = batchmappb.NewBatchMapClient(conn)
+	return c, nil
+}
+
+func NewFromClient(c batchmappb.BatchMapClient) (Client, error) {
+	return &client{
+		grpcClt: c,
+	}, nil
+}
+
+// CloseConn closes the grpc client connection.
+func (c *client) CloseConn(ctx context.Context) error {
+	return c.conn.Close()
+}
+
+// IsReady returns true if the grpc connection is ready to use.
+func (c *client) IsReady(ctx context.Context, in *emptypb.Empty) (bool, error) {
+	resp, err := c.grpcClt.IsReady(ctx, in)
+	if err != nil {
+		return false, err
+	}
+	return resp.GetReady(), nil
+}
+
+// BatchMapFn is the handler for the gRPC client (Numa container)
+// It takes in a stream of input Requests, sends them to the gRPC server(UDF) and then streams the
+// responses received back on a channel asynchronously.
+// We spawn 2 goroutines here, one for sending the requests over the stream
+// and the other one for receiving the responses
+func (c *client) BatchMapFn(ctx context.Context, inputCh <-chan *batchmappb.BatchMapRequest) (<-chan *batchmappb.BatchMapResponse, <-chan error) {
+	// errCh is used to track and propagate any errors that might occur during the rpc lifecyle, these could include
+	// errors in sending, UDF errors etc
+	// These are propagated to the applier for further handling
+	errCh := make(chan error)
+
+	// response channel for streaming back the results received from the gRPC server
+	// TODO(map-batch): Should we keep try to keep this buffered?
+	responseCh := make(chan *batchmappb.BatchMapResponse)
+
+	// BatchMapFn is a bidirectional streaming RPC
+	// We get a Map_BatchMapFnClient object over which we can send the requests,
+	// receive the responses asynchronously.
+	// TODO(map-batch): this creates a new gRPC stream for every batch,
+	// it might be useful to see the performance difference between this approach
+	// and a long-running RPC
+	stream, err := c.grpcClt.BatchMapFn(ctx)
+	if err != nil {
+		go func() {
+			errCh <- sdkerr.ToUDFErr("c.grpcClt.BatchMapFn stream", err)
+		}()
+		return responseCh, errCh
+	}
+
+	// read the response from the server stream and send it to responseCh channel
+	// any error is sent to errCh channel
+	go func() {
+		// close this channel to indicate that no more elements left to receive from grpc
+		// We do defer here on the whole go-routine as even during a error scenario, we
+		// want to close the channel and stop forwarding any more responses from the UDF
+		// as we would be replaying the current ones.
+		defer close(responseCh)
+
+		var resp *batchmappb.BatchMapResponse
+		var recvErr error
+		index := 0
+		for {
+			select {
+			// handle a context done
+			case <-ctx.Done():
+				errCh <- ctx.Err()
+				return
+			default:
+				resp, recvErr = stream.Recv()
+				// check if this is EOF error, which indicates that no more responses left to process on the
```

Comment:
Do we need a `ctx.Done()` check here? Since we create the gRPC stream with `ctx`, it seems to me that gRPC should handle it by throwing an error in the Recv() method if the `ctx` is cancelled. If we don't have a `ctx` check while closing the stream, there is a possibility that we could invoke `closeSend` to signal EOF to the gRPC stream while the `ctx` is cancelled. In this case, gRPC might return the actual result, but since we don't `recv` when there is a `ctx.Done()` signal, we might end up not receiving the result. This whole context setup is quite confusing. My suggestion is to rely on gRPC to handle `ctx` since we pass it while creating the stream. I believe we should adopt this approach in the reduce and ud source as well.

Reply:
I have updated the implementation
1) On the sending side, check for context.Done, and return without stream.CloseSend
2) On the recv side, the grpc server should send a ctx error back to the client.